Move metadata to pyproject.toml

cdce8p opened the pull request:
## What do these changes do?

The metadata changes:

```diff
 Metadata-Version: 2.1
 Name: aiohttp
 Version: 4.0.0a2.dev0
 Summary: Async http client/server framework (asyncio)
-Maintainer: aiohttp team <team@aiohttp.org>
-Maintainer-email: team@aiohttp.org
+Maintainer-email: aiohttp team <team@aiohttp.org>
 License: Apache-2.0
-Home-page: https://github.com/aio-libs/aiohttp
+Project-URL: Homepage, https://github.com/aio-libs/aiohttp
 Project-URL: Chat: Matrix, https://matrix.to/#/#aio-libs:matrix.org
 Project-URL: Chat: Matrix Space, https://matrix.to/#/#aio-libs-space:matrix.org
 Project-URL: CI: GitHub Actions, https://github.com/aio-libs/aiohttp/actions?query=workflow%3ACI
 Project-URL: Coverage: codecov, https://codecov.io/github/aio-libs/aiohttp
 Project-URL: Docs: Changelog, https://docs.aiohttp.org/en/stable/changes.html
 Project-URL: Docs: RTD, https://docs.aiohttp.org
 Project-URL: GitHub: issues, https://github.com/aio-libs/aiohttp/issues
 Project-URL: GitHub: repo, https://github.com/aio-libs/aiohttp
 Classifier: Development Status :: 5 - Production/Stable
 Classifier: Framework :: AsyncIO
 Classifier: Intended Audience :: Developers
 Classifier: License :: OSI Approved :: Apache Software License
 Classifier: Operating System :: POSIX
 Classifier: Operating System :: MacOS :: MacOS X
 Classifier: Operating System :: Microsoft :: Windows
 Classifier: Programming Language :: Python
 Classifier: Programming Language :: Python :: 3
 Classifier: Programming Language :: Python :: 3.9
 Classifier: Programming Language :: Python :: 3.10
 Classifier: Programming Language :: Python :: 3.11
 Classifier: Programming Language :: Python :: 3.12
 Classifier: Programming Language :: Python :: 3.13
 Classifier: Topic :: Internet :: WWW/HTTP
 Requires-Python: >=3.9
 Description-Content-Type: text/x-rst
 License-File: LICENSE.txt
 Requires-Dist: aiohappyeyeballs >=2.3.0
 Requires-Dist: aiosignal >=1.1.2
 Requires-Dist: frozenlist >=1.1.1
 Requires-Dist: multidict <7.0,>=4.5
 Requires-Dist: propcache >=0.2.0
 Requires-Dist: yarl <2.0,>=1.17.0
 Requires-Dist: async-timeout <6.0,>=4.0 ; python_version < "3.11"
 Provides-Extra: speedups
 Requires-Dist: brotlicffi ; (platform_python_implementation != "CPython") and extra == 'speedups'
 Requires-Dist: Brotli ; (platform_python_implementation == "CPython") and extra == 'speedups'
 Requires-Dist: aiodns >=3.2.0 ; (sys_platform == "linux" or sys_platform == "darwin") and extra == 'speedups'

 ...
```

**Open questions**
- Should this PR be backported? It's possible, might just require some manual work due to merge conflicts.

webknjaz commented:
Let's not rush this. Keeping a dedicated config is a conscious choice. Also, PRs must not reformat half a project along with changing the packaging configs.

FWIW I'd rather keep it as is for now. Especially, since I was going to sync the configs across the projects and this is going to cause a lot of headache since it conflicts with how things are now.

I'm 👎 on this migration right now. It'd be interesting to gather the current dependency limitations in the downstreams to make an informed choice. Additionally, the move would have to happen in the dependencies first.

asvetlov reviewed (commented):
I personally don't care where the project metadata sits until it works and I personally shouldn't spend additional time on the infrastructure maintenance.

asvetlov commented:
> That's actually a bug with the current black config.

@cdce8p could you please extract a fix for this particular problem into a separate PR? A small patch has a great chance to  be reviewed and approved quickly.

asvetlov commented:
Ooh, sorry. I've pushed the wrong button and closed the PR. Please accept my apologizes.

cdce8p commented:
> > That's actually a bug with the current black config.
> 
> @cdce8p could you please extract a fix for this particular problem into a separate PR? A small patch has a great chance to be reviewed and approved quickly.

See #9962. It looks bigger than it actually is.

asvetlov commented:
Thanks, @cdce8p 
The PR is merged and back ported

cdce8p commented:
Rebased the PR. Also moved the `dependencies` and `optional-dependencies` back to `setup.cfg` to purely focus on the static metadata. Please let me know if you're interested in moving this forward. If not or not at this time, feel free to close it.

Dreamsorcerer commented on setup.cfg:
```diff
@@ -1,55 +1,4 @@
-zip_safe = False
-include_package_data = True
-
 install_requires =
```
I would have thought the dependencies would go in the same file as the rest of the metadata..?

cdce8p replied:
Yes they can be moved as well. Just the current tooling expects them to be in `setup.cfg`. Since this PR stalled for so long I figured I'd revert that part for now. See my last commit. https://github.com/aio-libs/aiohttp/pull/9951/commits/ec9022547a07f710a83a5c3e172884ccc82632bb

_I'd be happy to re-include it here or do it in a followup if there is interest._

Dreamsorcerer replied:
Right. I'll defer to @webknjaz here to decide how it should go.

asvetlov commented on pyproject.toml:
```diff
@@ -1,9 +1,57 @@
+requires = ["setuptools >= 70.0.0"]
 build-backend = "setuptools.build_meta"
 
+[project]
```
Does dependabot understand this format?
I know that dependabot for poetry still requires the old `[tool.poetry]` section instead of the recommended `[project]`, I wonder if the setuptools-based check is different.

webknjaz replied:
I don't think dependabot updates abstract deps anyway. Renovate probably would. Though, in general, this shouldn't be updated automatically. It's not a lock file and mustn't pin things or contain top boundaries in general. This is something that pip-tools constraints would solve, as they'd lock envs w/o the dangers of causing problems in dependency resolution for the end users.

cdce8p replied:
As mentioned earlier, I reverted the part that moved the dependencies to only focus on the static metadata here.

webknjaz replied:
Ah, I didn't notice that you marked those as dynamic.

webknjaz replied:
But still, there's no need to separate those anyway. So let's get them back into this config.

webknjaz replied:
Ack. Follow-up is good. Marking the thread as resolved.

webknjaz commented:
> Rebased the PR. Also moved the `dependencies` and `optional-dependencies` back to `setup.cfg` to purely focus on the static metadata. Please let me know if you're interested in moving this forward. If not or not at this time, feel free to close it.

Not sure if this would work well with `setuptools`. I think when it detects PEP 621 mode, it would only source the spec-compliant bits from `pyproject.toml`. There might be some backend configs that would be sourced from other places, but anything that's not listed as dynamic explicitly is likely to be disregarded or otherwise cause breakage. So with this move we'll need to migrate things atomically.

webknjaz commented on pyproject.toml:
```diff
@@ -1,9 +1,57 @@
+version = {attr = "aiohttp.__version__"}
+
+[tool.setuptools.packages.find]
+include = ["aiohttp*"]
```
This will probably happen automatically.

webknjaz replied:
Go with the second code suggestion, then. I don't like that `aiohttp*` would match something that may get in accidentally, like `aiohttp1`.

webknjaz commented on setup.cfg:
```diff
@@ -1,55 +1,4 @@
-  aiohttp
-  aiohttp._websocket
-# https://setuptools.readthedocs.io/en/latest/setuptools.html#setting-the-zip-safe-flag
-zip_safe = False
```
This is no longer necessary?

webknjaz replied:
Ack. Marking the thread as resolved.

webknjaz commented on pyproject.toml:
```diff
@@ -1,9 +1,57 @@
+include = ["aiohttp*"]
+exclude = ["examples"]
+
+[tool.setuptools.package-data]
+"*" = ["*.so"]
+
+[tool.setuptools.exclude-package-data]
+"*" = ["*.c", "*.h"]
```
I'm not entirely sure if these are still needed. Might need some manual checks.

webknjaz commented on pyproject.toml:
```diff
@@ -1,9 +1,57 @@
-requires = [
-    "setuptools >= 46.4.0",
-]
+requires = ["setuptools >= 70.0.0"]
```
What's the specific feature that requires at least v70?

webknjaz commented:
> > I'm 👎 on this migration right now. It'd be interesting to gather the current dependency limitations in the downstreams to make an informed choice. Additionally, the move would have to happen in the dependencies first.
> 
> I'm probably missing something here. This PR doesn't affect any downstream projects. The project metadata change is minimal and spec compliant.
> 
> Tbh I just thought it might be a good time to move the static metadata after I learned that it can work fine with `setup.py`. It will probably still take a while (couple of years) but I expect setuptools to eventually deprecate the `setup.cfg` config.

What I meant back then was that I'd like the move to first happen in projects that are direct deps of aiohttp (like yarl). This is because I've implemented some advanced packaging techniques there (like an in-tree build backend). I'm going to eventually migrate aiohttp to use the same model. So packaging updates would flow from those places where improvements were first implemented.

Upon reflection, thought, I think that this PR doesn't block that so you're fine. Talking about the downstreams, they've been historically behind some upstream packaging standards, but AFAIK they've mostly implemented their own PEP 517 helpers by now. Still, there might be problems where the setuptools version they ship is older than our spec (it'd conflict, forcing them to patch it out on their side). Also, if they patch it out, there's a possibility that it's older than PEP 621 effectively not producing most of the metadata. So we should at least be careful about https://github.com/aio-libs/aiohttp/pull/9951/files#r2066217591.

cdce8p commented:
That's for the review @webknjaz! I'd address all comments in one to keep it readable. Please let me know if you have any followup questions.

> Move dependencies to `pyproject.toml`

Sounds good. I'd just recommend to do that in a followup PR. Otherwise the backports PRs will likely not apply cleanly. In the end it's just a revert of https://github.com/aio-libs/aiohttp/pull/9951/commits/ec9022547a07f710a83a5c3e172884ccc82632bb

With `dependencies` and `optional-dependencies` marked as `dynamic`, they are still read from `setup.cfg` as well. Checked the generated `METADATA` file again just to be sure.

> ```toml
> [tool.setuptools.packages.find]
> include = ["aiohttp*"]
> ```
> This will probably happen automatically.

Unfortunately not. Setuptools will throw an error `Multiple top-level packages discovered in a flat-layout: ['vendor', 'aiohttp', 'CHANGES']`. We can get rid of `exclude = ["examples"]` though.

> ```ini
> zip_safe = False
> ```
> This is no longer necessary?

Yes. From the setuptools docs
> Obsolete - only relevant for pkg_resources, easy_install and setup.py install in the context of eggs (deprecated).

https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#setuptools-specific-configuration

> > `package-data` and `exclude-package-data`
>
> I'm not entirely sure if these are still needed. Might need some manual checks.

`package-data` can indeed be removed. The default for `pyproject.toml` configs is `include-package-data = true`. We still need `exclude-package-data` though. Otherwise the wheels would include `*.c` and `*.h` files. I plan to do a followup to optimize the sdist / wheel sizes a bit by extending the list further. E.g. we still include the raw cython `*.pyx` files in the wheels which probably aren't necessary.

> What's the specific feature that requires at least v70?

Double checked that. It's actually `>= 66.1` that's required, so I'd suggest `>= 67.0`. Previous versions relied on `pkgutil.ImpImporter` which isn't available on Python 3.12+.

https://setuptools.pypa.io/en/latest/history.html#v66-1-0

> Still, there might be problems where the setuptools version they ship is older than our spec

Yes, I've seen a few reports in the light of the PEP 639 adoption. Support for it was only added in `v77`. I believe most (?) distributors still use an older version with no build isolation so that caused some confusion. That's one of the reasons I haven't updated the metadata here yet. `v67.0` should be fine.

cdce8p commented:
Pushed a new commit to incorporate the license metadata changes from #11226.

@webknjaz Would you mind taking a look at this again? I believe I've addressed all your comments in https://github.com/aio-libs/aiohttp/pull/9951#issuecomment-2841603641. Please let me know if you have any more questions.

webknjaz commented on MANIFEST.in:
I'd rather not make unrelated changes. The change note talks about moving the metadata definitions, but the patch has some random changes without any justification. Let's not do this.

webknjaz replied:
Marking this thread as resolved.

webknjaz commented on setup.cfg:
```diff
@@ -1,55 +1,4 @@
-long_description_content_type = text/x-rst
-maintainer = aiohttp team <team@aiohttp.org>
-maintainer_email = team@aiohttp.org
-license = Apache-2.0 AND MIT
```
This SPDX identifier got lost somewhere.

webknjaz replied:
@cdce8p have you re-checked if this affects the metadata after all the updates applied recently?

webknjaz commented on pyproject.toml:
```diff
@@ -1,10 +1,62 @@
 
+[project]
+name        = "aiohttp"
+license     = {text = "Apache-2.0 AND MIT"}
```
This is incorrect. SPDX must not be added as license text: https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license

webknjaz replied:
Okay. Marking this chain as resolved.

cdce8p commented:
> > I believe I've addressed all your comments in [#9951 (comment)](https://github.com/aio-libs/aiohttp/pull/9951#issuecomment-2841603641). Please let me know if you have any more questions.
> 
> All other threads need to be addressed as well. There's a number of those with zero response.

I've addressed them in https://github.com/aio-libs/aiohttp/pull/9951#issuecomment-2841603641 as I personally find the Github code comments hard to follow when it's a discussion about one or more related topics.

webknjaz commented on pyproject.toml:
```diff
@@ -1,10 +1,62 @@
+  "Programming Language :: Python :: 3.13",
+  "Topic :: Internet :: WWW/HTTP",
+]
+requires-python = ">=3.9"
```
Keep this sparse
```suggestion
requires-python = ">= 3.9"
```

webknjaz commented on pyproject.toml:
```diff
@@ -1,10 +1,62 @@
 requires = [
     "pkgconfig",
-    "setuptools >= 46.4.0",
+    "setuptools >= 67.0",
```
Marking this thread as resolved.

webknjaz commented:
@cdce8p short of a few cosmetic requests, this seems almost ready.

cdce8p commented:
@webknjaz Pushed another update to address the style changes and also responded to your questions. Please let me know if I missed anything.
